docs/proposals/ingress-ha-scheduling-provisioning.md

```diff
@@ -0,0 +1,194 @@
+  the IP addresss of the node where Ingress Pod is running. In case of a 
+  failure the Ingress Pod can be be moved to a different node.
+* How many Ingress Pod should run in a cluster? Should all Ingress Pod 
+  list&watch all Ingress Resource with out distinction? There is no way 
+  to bind or schedule ingress resource to a Ingress Pod/ReplicaSet(or a 
+  set of Ingress Pods/ReplicaSets), result in insufficient or excessive 
+  use of resource.
+
+## Goal
+This Proposal aims to address the above issues by the following mechanism:
+
+* Ingress HA: using keepalived and VIP to provide High Availability(mainly
+  for nginx/haproxy implementation, cloud implementation usually already 
+  provide HA)
+* Ingress Scheduling: schedule Ingress Resource to Ingress Pod/ReplicaSet
+* Ingress Provisioning: allow user to dynamically add Ingress Pod/ReplicaSet 
+  on demand
+
+## NonGoal
+
+* Ingress ReplicaSet rolling update
+
+## Ingress HA
+(AKA: Ingress Virtual IP using keepalived)
+
+#### High level design
+* use keepalived to provide HA
+* cluster admin choose a group of nodes which could be accessed out of cluster 
+  and are in the same L2 broacast domain to run Ingress Pod
+* deploy Ingress Pod using ReplicaSet(at least 2 replicas for HA)
+* using AntiAffinity feature so that Ingress Pod created by the same Ingress 
+  ReplicaSet could be scheduled to different node
+* cluster admin choose a CIDR for Ingress VIP(AKA IngreeVIPCIDR)
+* each Ingress Replicaset will be allocated a VIP from IngreeVIPCIDR(allocated by
+  cluster admin or API server)
+* Ingress Pods use host network
+* Ingress Pods created by the same Ingress ReplicaSet will run keepalived, only
+  one Ingress Pod will get the VIP
+* users out of cluster access incluster service by the Ingress VIP 
+
+#### Why VIP instead of round-robin DNS
+A question that pops up every now and then is why we do all this stuff 
+with virtual IPs rather than just use standard round-robin DNS. 
+There are a few reasons:
+
+* There is a long history of DNS libraries not respecting DNS TTLs and 
+    caching the results of name lookups.
+* Many apps do DNS lookups once and cache the results.
+* Even if apps and libraries did proper re-resolution, the load of every 
+    client re-resolving DNS over and over would be difficult to manage.
+
+#### Challenge
+* VIP is boud to Ingress ReplicaSet, how to expost it to Ingress Pod? One
+  approache is using ConfigMap, but then cluster admin need allocate VIP and 
+  write it to ConnfigMap, which makes automatic deployment harder. 
+* All Ingress Pods created by the same Ingress ReplicaSet need know others' RIP.
+
+## Ingress Scheduling
+
+#### High level design
+
+* Ingress ReplicaSets are created by cluster admin in advance
+* If all Ingress Pods are saturated, it's cluster admin's duty to create 
+  more Ingress ReplicaSets
+* There is a Ingress Scheduler which will schedule Ingress Resources to Ingress 
```

Comment:
IMO this might not be necessary. Users pick an Ingress claim based on QoS needs. The Ingress claim has a vip. The vip is backed by pods. If the pods are saturated, the admin or an autoscaler needs to scale them, but the vip doesn't change.